File: civil3d/plugin/src/AcadAdapter.cs
// Reads the active drawing into the Core snapshot: paper-space layouts =
// sheets, title-block attributes matched to rule labels, TEXT/MTEXT =
// text notes. Read-only.
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using Autodesk.AutoCAD.DatabaseServices;
using ArxChecker.Core;

namespace ArxChecker.Acad;

public static class Adapter
{
    private static readonly Regex NormalizeRe = new("[^A-Z0-9]+");

    private static string Normalize(string label) =>
        NormalizeRe.Replace((label ?? "").ToUpperInvariant(), "");

    public static (bool Found, string Value) MatchAttribute(
        Dictionary<string, string> attributes, string label, JsonObject paramMap)
    {
        var mapped = RulesStore.Str(paramMap, label);
        if (!string.IsNullOrEmpty(mapped))
        {
            foreach (var pair in attributes)
                if (string.Equals(pair.Key, mapped, StringComparison.OrdinalIgnoreCase))
                    return (true, pair.Value);
            return (false, null);
        }

        var wanted = Normalize(label);
        if (wanted.Length == 0) return (false, null);
        var normalized = new Dictionary<string, string>();
        foreach (var pair in attributes)
            normalized.TryAdd(Normalize(pair.Key), pair.Value);
        if (normalized.TryGetValue(wanted, out var exact)) return (true, exact);
        foreach (var pair in normalized)
            if (pair.Key.Contains(wanted) || (pair.Key.Length >= 3 && wanted.Contains(pair.Key)))
                return (true, pair.Value);
        return (false, null);
    }

    private const int MaxBlockDepth = 2;

    private static void CollectTexts(Transaction tr, BlockTableRecord btr, List<TextItem> texts,
                                     Dictionary<string, string> attributes, string contextSuffix,
                                     string zoomHandle, int depth)
    {
        foreach (ObjectId id in btr)
        {
            void Add(string text, string context)
            {
                if (!string.IsNullOrWhiteSpace(text))
                    texts.Add(new TextItem
                    {
                        Handle = id.Handle.ToString(),
                        ZoomHandle = zoomHandle,
                        Text = text.Trim(),
                        Context = context + contextSuffix,
                    });
            }

            switch (tr.GetObject(id, OpenMode.ForRead))
            {
                case DBText text:
                    Add(text.TextString, "text note");
                    break;
                case MText mtext:
                    // MText.Text is the plain text with inline codes stripped
                    Add(mtext.Text, "text note");
                    break;
                case MLeader mleader:
                    try
                    {
                        if (mleader.ContentType == ContentType.MTextContent && mleader.MText != null)
                            Add(mleader.MText.Text, "text note (leader)");
                    }
                    catch { /* block-content leaders have no MText */ }
                    break;
                case Dimension dimension:
                    var dimText = dimension.DimensionText;
                    if (!string.IsNullOrWhiteSpace(dimText) && dimText.Trim() != "<>")
                        Add(dimText.Replace("<>", " "), "text note (dimension override)");
                    break;
                case Table table:
                    for (int row = 0; row < table.Rows.Count; row++)
                        for (int col = 0; col < table.Columns.Count; col++)
                        {
                            try { Add(table.Cells[row, col].TextString, "text note in table"); }
                            catch { /* merged / non-text cells */ }
                        }
                    break;
                case BlockReference reference:
                    foreach (ObjectId attId in reference.AttributeCollection)
                    {
                        if (tr.GetObject(attId, OpenMode.ForRead) is not AttributeReference attribute
                            || string.IsNullOrEmpty(attribute.Tag))
                            continue;
                        attributes?.TryAdd(attribute.Tag, attribute.TextString);
                        // attribute values are searchable/replaceable per insert
                        if (!string.IsNullOrWhiteSpace(attribute.TextString))
                            texts.Add(new TextItem
                            {
                                Handle = attId.Handle.ToString(),
                                Text = attribute.TextString.Trim(),
                                Context = $"attribute \"{attribute.Tag}\"" + contextSuffix,
                            });
                    }
                    if (depth < MaxBlockDepth
                        && tr.GetObject(reference.BlockTableRecord, OpenMode.ForRead)
                            is BlockTableRecord definition
                        && !definition.IsFromExternalReference && !definition.IsLayout
                        && !definition.Name.StartsWith("*"))
                    {
                        // text inside the block definition; zoom targets the insert,
                        // and the "in block" context excludes it from find & replace
                        // (editing the definition would change every insert)
                        CollectTexts(tr, definition, texts, null,
                                     $" in block \"{definition.Name}\"",
                                     id.Handle.ToString(), depth + 1);
                    }
                    break;
            }
        }
    }

    public static Snapshot BuildSnapshot(Database db, string docName, JsonObject rules,
                                         bool includeModelSpace = true)
    {
        var paramMap = (rules["revit"] as JsonObject)?["paramMap"] as JsonObject;
        var fieldRules = RulesStore.EnabledRules(rules, "titleBlock", "revision").ToList();
        var dwgRule = fieldRules.FirstOrDefault(r => RulesStore.Str(r, "id") == "dwgNo");
        var otherRules = fieldRules.Where(r => RulesStore.Str(r, "id") != "dwgNo").ToList();

        var snapshot = new Snapshot { DocTitle = docName };

        using var tr = db.TransactionManager.StartTransaction();
        var layoutDict = (DBDictionary)tr.GetObject(db.LayoutDictionaryId, OpenMode.ForRead);
        var layouts = new List<Layout>();
        foreach (DBDictionaryEntry entry in layoutDict)
        {
            var layout = (Layout)tr.GetObject(entry.Value, OpenMode.ForRead);
            if (!layout.ModelType) layouts.Add(layout);
        }
        layouts.Sort((a, b) => a.TabOrder.CompareTo(b.TabOrder));

        foreach (var layout in layouts)
        {
            var texts = new List<TextItem>();
            var attributes = new Dictionary<string, string>();
            var btr = (BlockTableRecord)tr.GetObject(layout.BlockTableRecordId, OpenMode.ForRead);
            CollectTexts(tr, btr, texts, attributes, "", null, 0);

            var number = layout.LayoutName;
            if (dwgRule != null)
            {
                var (found, value) = MatchAttribute(attributes, RulesStore.Str(dwgRule, "label"), paramMap);
                if (found && !string.IsNullOrWhiteSpace(value)) number = value.Trim();
            }

            var sheet = new SheetData
            {
                LayoutName = layout.LayoutName,
                Number = number,
                Name = layout.LayoutName,
                TextNotes = texts,
            };
            foreach (var rule in otherRules)
            {
                var ruleId = RulesStore.Str(rule, "id");
                var (found, value) = MatchAttribute(attributes, RulesStore.Str(rule, "label"), paramMap);
                if (found) sheet.Params[ruleId] = value;
                else sheet.MissingParams.Add(ruleId);
            }

            foreach (var field in (JsonArray)rules["project"])
            {
                var id = RulesStore.Str(field, "id");
                if (snapshot.ProjectInfo.TryGetValue(id, out var existing) && !string.IsNullOrEmpty(existing))
                    continue;
                var (found, value) = MatchAttribute(attributes, RulesStore.Str(field, "label"), paramMap);
                if (found && !string.IsNullOrWhiteSpace(value))
                    snapshot.ProjectInfo[id] = value.Trim();
            }

            snapshot.Sheets.Add(sheet);
        }

        if (includeModelSpace)
        {
            var blockTable = (BlockTable)tr.GetObject(db.BlockTableId, OpenMode.ForRead);
            var modelBtr = (BlockTableRecord)tr.GetObject(
                blockTable[BlockTableRecord.ModelSpace], OpenMode.ForRead);
            var modelTexts = new List<TextItem>();
            CollectTexts(tr, modelBtr, modelTexts, null, "", null, 0);
            if (modelTexts.Count > 0)
                snapshot.Sheets.Add(new SheetData
                {
                    LayoutName = "Model", Number = "Model", Name = "Model",
                    IsModelSpace = true, TextNotes = modelTexts,
                });
        }

        tr.Commit();
        return snapshot;
    }

    public static string LayoutForPage(Snapshot snapshot, string page) =>
        snapshot.Sheets.FirstOrDefault(s => s.Number == page)?.LayoutName;
}

File: civil3d/checker/acad_adapter.py
# -*- coding: utf-8 -*-
"""COM adapter: reads the open Civil 3D / AutoCAD drawing into the same
model-snapshot shape the Revit adapter produces, so the shared rules
engine and spell checker run unchanged.

Mapping: paper-space layouts = sheets; title block = the attributed
block reference in each layout (attributes matched to rule labels);
DBText/MText entities = text notes. elementId carries the entity handle
(a string — the engine treats it as opaque). Each sheet dict also gets a
'layoutName' extra so actions can activate the right layout tab.
"""
import re

from .mtext import strip_mtext

_NORMALIZE_RE = re.compile(r'[^A-Z0-9]+')


def connect():
    """Attach to the running Civil 3D / AutoCAD session."""
    try:
        import win32com.client
    except ImportError:
        raise RuntimeError(
            'pywin32 is not installed. Run the launcher .bat once, or:\n'
            '  py -3 -m pip install pywin32')
    try:
        app = win32com.client.GetActiveObject('AutoCAD.Application')
    except Exception:
        raise RuntimeError(
            'Could not connect to Civil 3D / AutoCAD.\n'
            'Start Civil 3D, open the drawing, then run the checker again.')
    try:
        doc = app.ActiveDocument
    except Exception:
        raise RuntimeError('Civil 3D is running but no drawing is open.')
    return app, doc


def _normalize(label):
    return _NORMALIZE_RE.sub('', (label or '').upper())


def _match_attribute(attributes, label, param_map):
    """Find the attribute value for a rule label.

    Order: explicit paramMap tag, exact normalized match (DWG NO ->
    DWG_NO), then containment either way (DRAWING NUMBER ~ DWGNUMBER).
    Returns (found, value).
    """
    mapped = (param_map or {}).get(label)
    if mapped:
        for tag, value in attributes.items():
            if tag.upper() == mapped.upper():
                return True, value
        return False, None

    wanted = _normalize(label)
    if not wanted:
        return False, None
    normalized = dict((_normalize(tag), value) for tag, value in attributes.items())
    if wanted in normalized:
        return True, normalized[wanted]
    for tag, value in normalized.items():
        if wanted in tag or (len(tag) >= 3 and tag in wanted):
            return True, value
    return False, None


def _iter_layouts(doc):
    layouts = []
    for i in range(doc.Layouts.Count):
        layout = doc.Layouts.Item(i)
        if not layout.ModelType:
            layouts.append(layout)
    layouts.sort(key=lambda l: l.TabOrder)
    return layouts


def _collect_layout(layout):
    """Text entities and title-block attributes of one layout."""
    texts = []
    attributes = {}
    block = layout.Block
    for j in range(block.Count):
        entity = block.Item(j)
        object_name = entity.ObjectName
        if object_name == 'AcDbText':
            value = entity.TextString
            if value and value.strip():
                texts.append({'elementId': str(entity.Handle), 'text': value.strip()})
        elif object_name == 'AcDbMText':
            value = strip_mtext(entity.TextString)
            if value and value.strip():
                texts.append({'elementId': str(entity.Handle), 'text': value.strip()})
        elif object_name == 'AcDbBlockReference':
            try:
                has_attributes = entity.HasAttributes
            except Exception:
                has_attributes = False
            if has_attributes:
                for attribute in entity.GetAttributes():
                    tag = attribute.TagString
                    if tag and tag not in attributes:
                        attributes[tag] = attribute.TextString
    return texts, attributes


def build_snapshot(doc, rules_config):
    revit_settings = rules_config.get('revit') or {}
    param_map = revit_settings.get('paramMap') or {}
    field_rules = [
        r for r in rules_config.get('rules', [])
        if r.get('enabled') and r.get('category') in ('titleBlock', 'revision')
    ]
    dwg_rule = next((r for r in field_rules if r['id'] == 'dwgNo'), None)
    other_rules = [r for r in field_rules if r['id'] != 'dwgNo']
    project_fields = rules_config.get('project') or []

    snapshot = {
        'docTitle': doc.Name,
        'projectInfo': {},
        'sheets': [],
    }
    project_info = {}

    for layout in _iter_layouts(doc):
        texts, attributes = _collect_layout(layout)

        number = layout.Name
        if dwg_rule is not None:
            found, value = _match_attribute(attributes, dwg_rule['label'], param_map)
            if found and value and value.strip():
                number = value.strip()

        params = {}
        missing = []
        for rule in other_rules:
            found, value = _match_attribute(attributes, rule['label'], param_map)
            if found:
                params[rule['id']] = value
            else:
                missing.append(rule['id'])

        for field in project_fields:
            if project_info.get(field['id']):
                continue
            found, value = _match_attribute(attributes, field['label'], param_map)
            if found and value and value.strip():
                project_info[field['id']] = value.strip()

        snapshot['sheets'].append({
            'elementId': None,
            'layoutName': layout.Name,
            'number': number,
            'name': layout.Name,
            'params': params,
            'missingParams': missing,
            'textNotes': texts,
            'placedViews': [],
            'schedules': [],
        })

    snapshot['projectInfo'] = project_info
    return snapshot


def layout_for_page(snapshot, page):
    for sheet in snapshot.get('sheets', []):
        if sheet.get('number') == page:
            return sheet.get('layoutName')
    return None
